engine/consensus/approvals/core.go

```diff
@@ -0,0 +1,392 @@
+		return nil
+	}
+
+	for _, collector := range collectors {
+		// let's check at least that candidate block is deep enough for emergency sealing.
+		if collector.BlockHeight+sealing.DefaultEmergencySealingThreshold < lastFinalizedHeight {
+			err := collector.CheckEmergencySealing(lastFinalizedHeight)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *approvalProcessingCore) processPendingApprovals(collector *AssignmentCollector) error {
+	predicate := func(approval *flow.ResultApproval) bool {
+		return approval.Body.ExecutionResultID == collector.ResultID
+	}
+
+	// filter cached approvals for concrete execution result
+	for _, approvalID := range c.approvalsCache.Ids() {
+		if approval := c.approvalsCache.TakeIf(approvalID, predicate); approval != nil {
+			err := collector.ProcessAssignment(approval)
+			if err != nil {
+				if engine.IsInvalidInputError(err) {
+					c.log.Debug().
+						Hex("result_id", collector.ResultID[:]).
+						Err(err).
+						Msgf("invalid approval with id %s", approval.ID())
+				} else {
+					return fmt.Errorf("could not process assignment: %w", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *approvalProcessingCore) allCollectors() []*AssignmentCollector {
+	collectors := make([]*AssignmentCollector, 0, len(c.collectors))
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	for _, collector := range c.collectors {
+		collectors = append(collectors, collector)
+	}
+	return collectors
+}
+
+func (c *approvalProcessingCore) getCollector(resultID flow.Identifier) *AssignmentCollector {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.collectors[resultID]
+}
+
+// getOrCreateCollector performs lazy initialization of AssignmentCollector using double checked locking
+// Returns, (AssignmentCollector, true or false whenever it was created, error)
+func (c *approvalProcessingCore) getOrCreateCollector(result *flow.ExecutionResult) (*AssignmentCollector, bool, error) {
+	resultID := result.ID()
+	// first let's check if we have a collector already
+	collector := c.getCollector(resultID)
+	if collector != nil {
+		return collector, false, nil
+	}
+
+	// fast check shows that there is no collector, need to create one
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// we need to check again, since it's possible that after checking for existing collector but before taking a lock
+	// new collector was created by concurrent goroutine
+	collector, ok := c.collectors[resultID]
+	if ok {
+		return collector, false, nil
+	}
+
+	collector, err := NewAssignmentCollector(result, c.state, c.assigner, c.seals, c.verifier, c.approvalConduit,
+		c.requestTracker, c.getCachedBlockHeight, c.requiredApprovalsForSealConstruction)
+	if err != nil {
+		return nil, false, fmt.Errorf("could not create assignment collector for %v: %w", resultID, err)
+	}
+	c.collectors[resultID] = collector
+	return collector, true, nil
+}
+
+func (c *approvalProcessingCore) eraseCollectors(resultIDs []flow.Identifier) {
```

Comment:
We erase collectors by the given result ids, which are from the seals in the finalized block's payload. But the question is, can we say after the erase, the remaining collectors are for unsealed results? 

If there is sealed result that is not being erased, then they will stay in the collector forever causing memory leak. 

I think this is currently possible to happen, just like the edge case I mentioned in [ProcessApproval](https://github.com/onflow/flow-go/pull/632/files#r621581270). The edge case is that it's possible the following event happens:

- ProcessIncorporatedResult for block at height 20
- the sealed block height is `19`
- checkBlockOutdated returns false, so this result is for a unsealed block
- but before calling `getOrCreateCollector`, the OnFinalizedBlock is called, which updates the sealed height to 20, and also triggers eraseCollectors, and update lastSealedHeight as well as lastCleanupHeight
- now `getOrCreateCollector` is called, it still assumes the result is for an unsealed block (which is not), and create a collector and stores it in `collectors`, but it stores a sealed collector in the map, which will never be removed.

Reply:
This function is mainly for happy path cleanup. 
For cases like you mentioned we have cleanup of stale collectors:
https://github.com/onflow/flow-go/pull/632/files#diff-a78139d36fc688b1e2a452a8d7bafceb5755d38b25310a36719f7bb2538ce8b2R374

as soon as new block is sealed we will cleanup old collectors by height so that shouldn't be a problem